libsrc/serialization.cpp

```diff
@@ -0,0 +1,214 @@
+#include <sstream>
+#include <string>
+
+#include "ismrmrd/serialization.h"
+
+namespace ISMRMRD {
+
+void serialize(const Acquisition &acq, WritableStream &ws) {
+    AcquisitionHeader ahead = acq.getHead();
+    ws.write(reinterpret_cast<const char *>(&ahead), sizeof(AcquisitionHeader));
```

Comment:
I might be paranoid here, but I would add static assertions that classes such as `AcquisitionHeader` that are writing out in this way have the exact layout that we are expecting. `is_standard_layout`, the overall `sizeof` and the `offsetof` each member variable.

Reply:
I think it would be nice, but should probably happen in the main Ismrmrd header? After all, it is a pretty important guarantee.

Reply:
I went through and added static_asserts in the `ismrmrd.h` and `waveform.h` headers. So we now have the checks. Unfortunately, I also made the discovery that the waveform header has an 8-byte boundary while the other structures use 2-byte boundaries. So that header is 40 bytes intead of 32 and the version field (which is a 16 byte field) actually takes up 8 bytes. One can debate which boundary is the best, I think it is a problem that it is inconsistent between the different headers. This 8-byte boundary is baked in in many places, so we can't change it now, but the documentation that @kspaceKelvin is working on does not reflect this discrepancy, so it should be updated to reflect this.

Reply:
@dchansen was it accidental that it ended up like that or?

Reply:
Yes - absolutely an accident. I didn't stop to consider the layout implications of the design when it was originally made. By the time I realised, it was already released and in use.

Reply:
We are checking now that the current layout doesn't change. Also added standard layout checks on the C++ classes that inherit from the base structs. All of that is wrapped in C++11 checks.